keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,1067 @@
+were implemented as a more constrained subset of this problem and addressed how
+to integrate the Kubernetes type system with CEL.
+
+## Considerations
+
+### Admission Webhook Parity
+
+Users of the Kubernetes API are already familiar with
+ValidatingWebhookConfigurations. We should strive for consistency with this API
+unless there is a good reason to diverge from it.  As a concrete example, we
+should provide access to all the information that webhooks have access to (see
+[AdmissionRequest](https://github.com/kubernetes/kubernetes/blob/2ac6a4121f5b2a94acc88d62c07d8ed1cd34ed63/staging/src/k8s.io/api/admission/v1/types.go#L40))
+and if we provide access to additional information we should extend
+AdmissionRequest to include it.
+
+### Configurability
+
+Consider an admission rule that disallows requests based on a blocklist.
+
+While it is possible to inline a blocklist directly into a CEL expression as a
+data literal (`!(object.metadata.name in ['blocked1', 'blocked2'])`), there are
+a couple problems:
+
+- Long blocklists become unwieldy quickly in CEL expressions
+- Having something like a different blocklist for each namespace gets really
+  messy
+
+The need to configure admission rules is common enough (see below use cases)
+that we propose configuration be a 1st class concept in the API.
+
+Since all the policy frameworks we have surveyed have configurability as a 1st
+class concept, omitting it would result in either the policy frameworks not
+adopting this enhancement (and sticking with webhooks) or somehow bypassing the
+limitation. One possible approach would be to generate a CEL expression with the
+configuration data embedded as a data literal, but we would strongly prefer not
+to have policy frameworks generating a CEL expression for each possible
+configuration.
+
+### Migration
+
+With webhooks already in large scale use in the Kubernetes ecosystem, we intend
+to prioritize capabilities that ease migration. As a concrete example, when
+migrating, having fine grained control of what validation messages are returned
+and how they are formatted can make a migration far more seamless.
+
+### Compliance
+
+In-process admission control has fundamental advantages over webhooks: it is far
+safer to use in a "fail closed" mode. With webhooks, using "fail closed" can
+negatively impact cluster availability. But "fail closed" is very valuable when
+enforcing compliance (and security). We intend to prioritize capabilities that
+make "fail closed" a safe mode of operation. As a concrete example, only
+allowing CEL expressions that pass compilation and type checking significantly
+reduces the opportunities for runtime errors.
+
+Also, making it possible (and convenient) to declare zero trust policies is
+important to compliance. i.e., it should be possible for policies governing 
+resources like namespace and roles to default to the most restrictive state
+when first created.
```

Comment:
This is a good list. I've threaded them into this KEP.  "static or on-initialization" I listed as a non-goal since is seems that it should be a generically added feature not specific to admission. The other's I've grouped in with some of the safety feature TODOs in phase 1 and 2. I'm not sure which ones we will commit to yet, but we shouldn't loose track of them.